### PR TITLE
Serialization instances for `ConwayCertPredFailure`

### LINKED ISSUE #3476: ConwayCertPredFailure serialization instances
We need instances for:
* DecCBOR (ConwayCertPredFailure era)
* EncCBOR (ConwayCertPredFailure era)

### PR DESCRIPTION
**Pull request by teodanciu**
# Description
Closes https://github.com/input-output-hk/cardano-ledger/issues/3476
# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Cert.hs [hunk 1]
```diff
@@ -20,15 +21,18 @@ module Cardano.Ledger.Conway.Rules.Cert (
...
 import Cardano.Ledger.Conway.Era (ConwayCERT, ConwayDELEG, ConwayPOOL, ConwayVDEL)
-import Cardano.Ledger.Conway.Rules.Deleg (ConwayDelegPredFailure)
+import Cardano.Ledger.Conway.Rules.Deleg (ConwayDelegPredFailure (..))
 import Cardano.Ledger.Conway.Rules.VDel (ConwayVDelPredFailure, VDelEnv (VDelEnv))
 import Cardano.Ledger.Conway.TxCert (ConwayCommitteeCert, ConwayDelegCert, ConwayTxCert (..))
 import Cardano.Ledger.Shelley.API (CertState (..), DState, DelegEnv (DelegEnv), DelplEnv (DelplEnv), PState, PoolEnv (PoolEnv), VState)
 import Cardano.Ledger.Shelley.Rules (ShelleyPoolPredFailure)
 import Control.DeepSeq (NFData)
 import Control.State.Transition.Extended (Embed, STS (..), TRC (TRC), TransitionRule, judgmentContext, trans, wrapEvent, wrapFailed)
+import Data.Data (Typeable)
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Cert.hs]:
```suggestion
import Data.Typeable (Typeable)
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs [hunk 2]
```diff
@@ -35,12 +38,24 @@ import Control.State.Transition (
   State,
   transitionRules,
  )
+import Data.Data (Typeable)
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs]:
```suggestion
import Data.Typeable (Typeable)
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs [hunk 3]
```diff
@@ -35,12 +38,24 @@ import Control.State.Transition (
...
+import Data.Word (Word8)
 import GHC.Generics (Generic)
 import NoThunks.Class (NoThunks (..))
 
-data ConwayVDelPredFailure era
+data ConwayVDelPredFailure era = ConwayVDelPredFailure
   deriving (Show, Eq, Generic, NoThunks, NFData)
 
+instance (Typeable era) => EncCBOR (ConwayVDelPredFailure era) where
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs]:
```suggestion
instance Typeable era => EncCBOR (ConwayVDelPredFailure era) where
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs [hunk 4]
```diff
@@ -35,12 +38,24 @@ import Control.State.Transition (
...
+  encCBOR = \case
+    ConwayVDelPredFailure -> encodeListLen 1 <> encCBOR (0 :: Word8)
+
+instance (Typeable era) => DecCBOR (ConwayVDelPredFailure era) where
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs]:
```suggestion
instance Typeable era => DecCBOR (ConwayVDelPredFailure era) where
```

**Review by lehins — APPROVED** [review]
Looks great!

**Review by JaredCorduan — APPROVED** [review]
wonderful, thank you!

**Pull request merged**